ui/text_displayer.js

```diff
@@ -169,130 +169,128 @@ shaka.ui.TextDisplayer = class {
-            if (currentValue) {
-              tokens.push({type: tagType, value: currentValue});
-            }
-          }
-          currentValue = '';
-        } else {
+        // Seek to end of tag.
+        i++;
+        let wellTerminated = false;
+        for (;i < str.length; i++) {
           currentValue += str[i];
+          if (str[i] === '>') {
+            wellTerminated = true;
+            break;
+          }
         }
-      }
 
-      // Final flush.
-      if (currentValue) {
-        tokens.push({type: 'text', value: currentValue});
+        if (!wellTerminated || currentValue.length < 3) {
+          // No termination found, append literal string.
+          tokens.push({type: 'text', value: currentValue});
+        } else {
+          let tagType = 'startTag';
+          currentValue = currentValue.substr(1, currentValue.length - 2);
+          if (currentValue[0] == '/') {
+            currentValue = currentValue.substr(1); // Removes /
+            tagType = 'endTag';
+          }
+          if (currentValue) {
+            tokens.push({type: tagType, value: currentValue});
+          }
+        }
+        currentValue = '';
+      } else {
+        currentValue += str[i];
       }
+    }
 
-      return tokens;
-    };
-
-    const tokens = tokenize(cueBody);
-    if (!tokens) {
-      return [];
+    // Final flush.
+    if (currentValue) {
+      tokens.push({type: 'text', value: currentValue});
     }
 
-    const buildElement = (tokens) => {
-      const supportedTags = [
-        'i',
-        'b',
-        'u',
-      ];
-
-      const elements = [];
-
-      // Build a set of root elements from the token collection.
-      const stack = [];
-      for (let i = 0; i < tokens.length; i++) {
-        const currentToken = tokens[i];
-        const supportedTag = supportedTags.includes(currentToken.value);
-        if (currentToken.type === 'text' || !supportedTag) {
-          // Element is a text field or an unsupported tag.
-          const textElement = document.createElement('span');
-          textElement.textContent = currentToken.value;
-          if (stack.length > 0) {
-            // Add to bottom stack element.
-            stack[stack.length - 1].appendChild(textElement);
-          } else {
-            // Add as standalone.
-            elements.push(textElement);
-          }
+    return tokens;
+  }
+
+  /**
+   * Returns a safe collection of html elements for display as a single cue.
+   *
+   * @param {!Array<{type: string, value: string}>} tokens
+   * @return {!Array<HTMLElement>}
+   * @private
+   */
+  buildCueElements_(tokens) {
+    const supportedTags = [
+      'i',
+      'b',
+      'u',
+    ];
+
+    const elements = [];
+
+    // Build a set of root elements from the token collection.
+    const stack = [];
+    for (const currentToken of tokens) {
+      const supportedTag = supportedTags.includes(currentToken.value);
+      if (currentToken.type === 'text' || !supportedTag) {
```

Comment:
It seems like this would really screw up on voice tags, turning something like this:

```
<v:john>Hello!</v:john>
```

Into this:

```
<span>v:john</span><span>Hello!</span><span>v:john</span>
```

I think you should ignore unsupported tags instead of treating them like text.  Then you could be generating this much more sensible thing:

```
<span>Hello!</span>
```

It looks like adding voice support or class support might not even be that tough after your PR, but for now, we should at least ignore markup we don't support yet.

Reply:
I initially thought of ignoring those tags too, but then decided to display them as an indication that _something_ is wrong. I'm still very on the fence about it, so I've deferred to your preference with 770fe76.